### PR TITLE
Redox OS support

### PR DESCRIPTION
**Pull request by jD91mZM2**
This was a part of my RSoC project of [porting tokio](https://www.redox-os.org/news/rsoc-porting-tokio-to-redox/) and I think it's complete now :)

Tested:
 - Two small custom examples
 - 11/13 tokio examples work (udp servers might not be working but I'm getting `0.0.0.0:0` straight from rust which indicates this is a problem with either my setup or rust itself)
 - 6/7 hyper examples work (`web_api` is trying to connect to itself, which I think is generally broken - you can't even communicate between two netcat instances running on redox)

### CODE REVIEW

**Comment by jD91mZM2** [comment]
Oof, seems like our net2 patch breaks windows support. Thank you AppVeyor :)

**Comment by jD91mZM2** [comment]
Sometimes I really wish cargo would let you apply a patch to all dependencies

**Comment by jD91mZM2** [comment]
Sorry for the delay - it's finally upstreamed :smile:

**Comment by jD91mZM2** [comment]
Initially I thought the version bumps were necessary to get miow to use the latest version of dependencies and therefore support redox. But Cargo is nice enough that `0.2.5` will also match `0.2.33` so a `cargo update` would suffice. I will separate it from this PR shortly.

Making the platform unstable for a while does not work very well with tokio since people would then indirectly use mio and can't control the features.

I guess I'll need to step up as maintainer for this.

**Comment by carllerche** [comment]
We can continue the details of the process (stable vs. unstable) here: #853

**Comment by jD91mZM2** [comment]
Bump?

File: src/sys/redox/awakener.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            Ok(Awakener {
+                reader: rd,
+                writer: wr,
+            })
+        }
+
+        pub fn wakeup(&self) -> io::Result<()> {
+            match (&self.writer).write(&[1]) {
+                Ok(_) => Ok(()),
+                Err(e) => {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/awakener.rs]: Just an opinion, might not need to change, but I prefer the following code:

```rust
match (&self.writer).write(&[1]) {
    Ok(_) => Ok(()),
    Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => Ok(()),
    Err(err) => Err(err),
}
```

File: src/sys/redox/awakener.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            loop {
+                // Consume data until all bytes are purged
+                match (&self.reader).read(&mut buf) {
+                    Ok(i) if i > 0 => {},
+                    _ => return,
+                }
+            }
+        }
+
+        fn reader(&self) -> &redox::Io {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/awakener.rs]: This can be removed I think (see comment at `Evented`)?

File: src/sys/redox/awakener.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            &self.reader
+        }
+    }
+
+    impl Evented for Awakener {
+        fn register(&self, poll: &Poll, token: Token, interest: Ready, opts: PollOpt) -> io::Result<()> {
+            self.reader().register(poll, token, interest, opts)
```
Comment (Thomasdezeeuw) [inline src/sys/redox/awakener.rs]: Why not use the `reader` field directly, same for other methods.

File: src/sys/redox/awakener.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                    if e.kind() == io::ErrorKind::WouldBlock {
+                        Ok(())
+                    } else {
+                        Err(e)
+                    }
+                }
+            }
+        }
+
+        pub fn cleanup(&self) {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/awakener.rs]: I don't see this being called anywhere. Can this be removed?

File: src/sys/redox/eventedfd.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
+use {io, poll, Evented, Ready, Poll, PollOpt, Token};
+use std::os::unix::io::RawFd;
+
+/*
+ *
+ * ===== EventedFd =====
+ *
+ */
+
+#[derive(Debug)]
```
Comment (Thomasdezeeuw) [inline src/sys/redox/eventedfd.rs]: Move this after the description, it's sort of floating around currently.

File: src/sys/redox/eventedfd.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+///     fn deregister(&self, poll: &Poll) -> io::Result<()> {
+///         EventedFd(&self.fd).deregister(poll)
+///     }
+/// }
+/// ```
+///
+/// [`Evented`]: ../event/trait.Evented.html
+/// [`Poll`]: ../struct.Poll.html
+/// [`Poll::register`]: ../struct.Poll.html#method.register
+pub struct EventedFd<'a>(pub &'a RawFd);
```
Comment (Thomasdezeeuw) [inline src/sys/redox/eventedfd.rs]: Could the type definition of this be shared with the unix module, having only differente `Evented` implementations? This way we don't have to keep the documentation in sync.

File: src/sys/redox/io.rs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+/// Manages a FD
+#[derive(Debug)]
+pub struct Io {
+    fd: RawFd,
+}
+
+impl FromRawFd for Io {
+    unsafe fn from_raw_fd(fd: RawFd) -> Io {
+        Io { fd: fd }
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: Can use `Io { fd }`.

File: src/sys/redox/io.rs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+use std::os::unix::io::{IntoRawFd, AsRawFd, FromRawFd, RawFd};
+use syscall::{fcntl, F_SETFL, F_GETFL, O_NONBLOCK};
+
+pub fn set_nonblock(fd: RawFd) -> io::Result<()> {
+    let flags = fcntl(fd, F_GETFL, 0).map_err(super::from_syscall_error)?;
+    fcntl(fd, F_SETFL, flags | O_NONBLOCK).map_err(super::from_syscall_error)
+                                                     .map(|_| ())
+}
+
+
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: Double empty line.

File: src/sys/redox/io.rs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+impl Read for Io {
+    fn read(&mut self, dst: &mut [u8]) -> io::Result<usize> {
+        <&Io as Read>::read(&mut &*self, dst)
+    }
+}
+
+impl<'a> Read for &'a Io {
+    fn read(&mut self, dst: &mut [u8]) -> io::Result<usize> {
+        use syscall::read;
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: Move this to the imports/uses at the top of the file.

File: src/sys/redox/io.rs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl<'a> Write for &'a Io {
+    fn write(&mut self, src: &[u8]) -> io::Result<usize> {
+        use syscall::write;
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: Move this to the imports/uses at the top of the file.

File: src/sys/redox/io.rs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    fn deregister(&self, poll: &Poll) -> io::Result<()> {
+        poll::selector(poll).deregister(self.fd)
+    }
+}
+
+impl Read for Io {
+    fn read(&mut self, dst: &mut [u8]) -> io::Result<usize> {
+        <&Io as Read>::read(&mut &*self, dst)
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: You can use `(&self).read(&mut dst)`, I think.

File: src/sys/redox/io.rs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        use syscall::read;
+
+        read(self.as_raw_fd(), dst)
+            .map_err(super::from_syscall_error)
+    }
+}
+
+impl Write for Io {
+    fn write(&mut self, src: &[u8]) -> io::Result<usize> {
+        <&Io as Write>::write(&mut &*self, src)
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: You can use `(&self).write(&src)`, I think.

File: src/sys/redox/io.rs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Drop for Io {
+    fn drop(&mut self) {
+        use syscall::close;
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: Move this to the imports/uses at the top of the file.

File: src/sys/redox/io.rs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        let _ = close(self.as_raw_fd());
```
Comment (Thomasdezeeuw) [inline src/sys/redox/io.rs]: Maybe add a note about why the error is ignored. Likely can be the same as in the unix module, or otherwise the once from the standard library.

File: src/sys/redox/mod.rs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+pub use self::awakener::Awakener;
+pub use self::eventedfd::EventedFd;
+pub use self::io::{Io, set_nonblock};
+pub use self::selector::{Events, Selector};
+pub use self::tcp::{TcpStream, TcpListener};
+pub use self::udp::UdpSocket;
+
+use std::os::unix::io::FromRawFd;
+
+pub fn pipe() -> ::io::Result<(Io, Io)> {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/mod.rs]: It's unclear to my which `Result` your using here, I assume it's `std::io::Result`, but starting with `::` I'm not sure.

File: src/sys/redox/mod.rs [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    use syscall::{O_NONBLOCK, O_CLOEXEC};
```
Comment (Thomasdezeeuw) [inline src/sys/redox/mod.rs]: Move the imports to the top of the file with the rest.

File: src/sys/redox/mod.rs [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    use syscall::pipe2;
+
+    let mut fds = [0; 2];
+    try!(pipe2(&mut fds, O_NONBLOCK | O_CLOEXEC).map_err(from_syscall_error));
```
Comment (Thomasdezeeuw) [inline src/sys/redox/mod.rs]: `?` is prefered.

File: src/sys/redox/mod.rs [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+pub use self::tcp::{TcpStream, TcpListener};
+pub use self::udp::UdpSocket;
+
+use std::os::unix::io::FromRawFd;
+
+pub fn pipe() -> ::io::Result<(Io, Io)> {
+    use syscall::{O_NONBLOCK, O_CLOEXEC};
+    use syscall::pipe2;
+
+    let mut fds = [0; 2];
```
Comment (Thomasdezeeuw) [inline src/sys/redox/mod.rs]: Maybe use `-1` so the code can't even mess with fd 0. Syscalls on fd `-1` should returns errors (right?).

File: src/sys/redox/mod.rs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    try!(pipe2(&mut fds, O_NONBLOCK | O_CLOEXEC).map_err(from_syscall_error));
+
+    unsafe {
+        Ok((Io::from_raw_fd(fds[0]), Io::from_raw_fd(fds[1])))
+    }
+}
+
+pub fn from_syscall_error(err: ::syscall::Error) -> ::io::Error {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/mod.rs]: Again starting with `::` is somewhat unclear to which type this is refering, same for the body of the code.

File: src/sys/redox/selector.rs [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+/// Each Selector has a globally unique(ish) ID associated with it. This ID
+/// gets tracked by `TcpStream`, `TcpListener`, etc... when they are first
+/// registered with the `Selector`. If a type that is previously associatd with
+/// a `Selector` attempts to register itself with a different `Selector`, the
+/// operation will return with an error. This matches windows behavior.
+static NEXT_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+
+#[derive(Debug)]
+pub struct Selector {
+    id: usize,
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: This doesn't seem to be used, can this be removed?

File: src/sys/redox/selector.rs [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Wait for events from the OS
+    pub fn select(&self, evts: &mut Events, awakener: Token, timeout: Option<Duration>) -> io::Result<bool> {
+        let mut timeout_fd = None;
+        if let Some(timeout) = timeout {
+            let mut file = File::open(format!("time:{}", CLOCK_MONOTONIC))?;
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: Do you want to allocate a String everytime select is called?

This could be resolved in a future pr.

File: src/sys/redox/selector.rs [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+
+            // TODO: use try_from below when stable
+            if timeout.as_secs() > ::std::i64::MAX as u64 {
+                panic!("too high duration");
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: Maybe mention it's above the timeout to `mio::Poll`, this message is too vague.

File: src/sys/redox/selector.rs [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+            }
+
+            let mut time = syscall::TimeSpec::default();
+            file.read(&mut time)?;
+
+            //tv_sec += i64::try_from(timeout.as_secs()).expect("too high duration"),
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: Remove?

File: src/sys/redox/selector.rs [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+    sync::{
+        Mutex,
+        atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT}
+    },
+    time::Duration
+};
+use syscall::{self, CLOCK_MONOTONIC, EVENT_READ, EVENT_WRITE};
+use {io, Ready, PollOpt, Token};
+
+const TIMEOUT_TOKEN: Token = Token(::std::usize::MAX - 1);
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: Documentated somewhere that this is not a valid token for use by the user of this crate.

File: src/sys/redox/selector.rs [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+            time.tv_nsec += timeout.subsec_nanos() as i32;
+
+            file.write(&time)?;
+
+            self.inner_register(file.as_raw_fd(), TIMEOUT_TOKEN, EVENT_READ)?;
+            timeout_fd = Some(file);
+        }
+
+        let cnt;
+        unsafe {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: You can let this unsafe block return `cnt`:

```rust
let cnt = unsafe {
    // ...
    cnt
};
```

File: src/sys/redox/selector.rs [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+            let bytes = (&self.efd).read(slice::from_raw_parts_mut(
+                evts.events.as_mut_ptr() as *mut u8,
+                evts.events.capacity() * mem::size_of::<syscall::Event>()
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: I think this is incorrect. You're going beyond the bounds of the slice.

File: src/sys/redox/selector.rs [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+/// registered with the `Selector`. If a type that is previously associatd with
+/// a `Selector` attempts to register itself with a different `Selector`, the
+/// operation will return with an error. This matches windows behavior.
+static NEXT_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+
+#[derive(Debug)]
+pub struct Selector {
+    id: usize,
+    efd: File,
+    tokens: Mutex<BTreeMap<RawFd, BTreeSet<Token>>>
```
Comment (Thomasdezeeuw) [inline src/sys/redox/selector.rs]: Why not use `HashMap` and `HashSet`? Just curious.

File: src/sys/redox/tcp.rs [hunk 28]
```diff
@@ -0,0 +1,285 @@
...
+    pub fn peer_addr(&self) -> io::Result<SocketAddr> {
+        self.inner.peer_addr()
+    }
+
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.inner.local_addr()
+    }
+
+    pub fn try_clone(&self) -> io::Result<TcpStream> {
+        self.inner.try_clone().map(|s| {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/tcp.rs]: `self.inner.try_clone().map(TcpStream::from_stream)`?

File: src/sys/redox/tcp.rs [hunk 29]
```diff
@@ -0,0 +1,285 @@
...
+            total += n;
+            if n < buf.len() {
+                return Ok(total);
+            }
+        }
+        Ok(total)
+    }
+}
+
+impl<'a> Read for &'a TcpStream {
```
Comment (Thomasdezeeuw) [inline src/sys/redox/tcp.rs]: As with other `Read`/`Write` implementations on references; this is ok to do in Redox?

File: src/sys/redox/udp.rs [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+
+    pub fn leave_multicast_v6(&self,
+                              multiaddr: &Ipv6Addr,
+                              interface: u32) -> io::Result<()> {
+        self.io.leave_multicast_v6(multiaddr, interface)
+    }
+
+    pub fn set_only_v6(&self, _only_v6: bool) -> io::Result<()> {
+        //self.io.set_only_v6(only_v6)
+        Err(io::Error::new(io::ErrorKind::Other, "Not implemented"))
```
Comment (Thomasdezeeuw) [inline src/sys/redox/udp.rs]: Maybe `unimplemented!()`? Since I would like to know this while developing, but I'm not sure about this.

File: src/deprecated/redox.rs [hunk 31]
```diff
@@ -0,0 +1,169 @@
...
+    }
+}
+
+impl Read for PipeReader {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.io.read(buf)
+    }
+}
+
+impl<'a> Read for &'a PipeReader {
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/deprecated/redox.rs]: I *think* it does. It has a SchemeMut trait and I'm sure that it locks that in a safe way

File: src/deprecated/redox.rs [hunk 32]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+#[derive(Debug)]
+pub struct PipeReader {
+    io: Io,
+}
+
+impl PipeReader {
+    pub fn from_stdout(stdout: process::ChildStdout) -> io::Result<Self> {
+        match sys::set_nonblock(stdout.as_raw_fd()) {
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/deprecated/redox.rs]: Looks like most of this file was a copy paste of the unix code. Perhaps that should be cleaned up as well :stuck_out_tongue:

File: src/deprecated/redox.rs [hunk 33]
```diff
@@ -0,0 +1,169 @@
...
+            Err(e) => return Err(e),
+            _ => {},
+        }
+        return Ok(PipeReader::from(unsafe { Io::from_raw_fd(stdout.into_raw_fd()) }));
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/deprecated/redox.rs]: Unix does this

File: src/net/tcp.rs [hunk 34]
```diff
@@ -7,12 +7,16 @@
 //!
 /// [portability guidelines]: ../struct.Poll.html#portability
 
+#[cfg(windows)]
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/net/tcp.rs]: No it doesn't? It compiles, and only windows uses SocketAddrV4

File: src/net/tcp.rs [hunk 35]
```diff
@@ -515,6 +528,7 @@ impl TcpListener {
     /// socket is desired then the `net2::TcpBuilder` methods can be used in
     /// combination with the `TcpListener::from_listener` method to transfer
     /// ownership into mio.
+    #[cfg(all(any(unix, windows), not(target_os = "redox")))]
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/net/tcp.rs]: I have no idea what went on in my head when I wrote that. :stuck_out_tongue:

File: src/sys/redox/awakener.rs [hunk 36]
```diff
@@ -0,0 +1,74 @@
+pub use self::pipe::Awakener;
+
+/// Default awakener backed by a pipe
+mod pipe {
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/sys/redox/awakener.rs]: This seems to be mostly copy pasted from the unix module

File: src/sys/redox/awakener.rs (hunk 4 above)
Comment (jD91mZM2) [reply src/sys/redox/awakener.rs]: `self.readiness_queue.inner.awakener.cleanup();`
src/poll.rs

File: src/sys/redox/eventedfd.rs (hunk 6 above)
Comment (jD91mZM2) [reply src/sys/redox/eventedfd.rs]: You're welcome to suggest where I'd place it

File: src/sys/redox/selector.rs (hunk 20 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: `selector_id != poll.selector.id()`
src/poll.rs

File: src/sys/redox/selector.rs (hunk 21 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: Good point, this is static and could be cached. Optimally would be if Rust would just allow `concat!`ing with constants :(

File: src/sys/redox/selector.rs (hunk 23 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: When it becomes stable I want to use this code instead.

File: src/sys/redox/selector.rs (hunk 24 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: Is it ever documented that you can't use the awakener, `usize::MAX`?

File: src/sys/redox/selector.rs (hunk 26 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: ???

I'm taking the full vector, including the uninitialized contents, as a byte slice.
Then depending on how much the operating system read I `set_len` to update the vector's length.

File: src/sys/redox/selector.rs (hunk 27 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: No idea, I just went with what Jeremy did. I heard someone mention you for some reason couldn't use HashMaps inside the redox kernel, so perhaps he just got used to BTrees...

File: src/sys/redox/udp.rs (hunk 30 above)
Comment (jD91mZM2) [reply src/sys/redox/udp.rs]: The idea is that an application can fall back to other error handling

File: src/net/tcp.rs [hunk 37]
```diff
@@ -538,12 +551,15 @@ impl TcpListener {
         })
     }
 
-    #[deprecated(since = "0.6.13", note = "use from_std instead")]
-    #[cfg(feature = "with-deprecated")]
-    #[doc(hidden)]
-    pub fn from_listener(listener: net::TcpListener, _: &SocketAddr)
```
[the remark this answers is not in the document]
Comment (jD91mZM2) [reply src/net/tcp.rs]: I have not got the slightest idea. I'm just gonna assume it was a problem with the rebase ¯\\\_(ツ)_/¯

File: src/sys/redox/selector.rs (hunk 21 above)
Comment (jD91mZM2) [reply src/sys/redox/selector.rs]: Although, fun fact, `File::open` allocates on each occurance on linux. This is because it needs a terminating nul byte. So this isn't that expensive

File: src/deprecated/redox.rs (hunk 31 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/deprecated/redox.rs]: Ok, then this looks good to me.

File: src/deprecated/redox.rs [hunk 38]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    fn deregister(&self, poll: &Poll) -> io::Result<()> {
+        self.io.deregister(poll)
+    }
+}
+
+impl From<Io> for PipeReader {
+    fn from(io: Io) -> PipeReader {
+        PipeReader { io: io }
```
Comment (Thomasdezeeuw) [inline src/deprecated/redox.rs]: `PipeReader { io }` works here.

File: src/deprecated/redox.rs [hunk 39]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    fn deregister(&self, poll: &Poll) -> io::Result<()> {
+        self.io.deregister(poll)
+    }
+}
+
+impl From<Io> for PipeWriter {
+    fn from(io: Io) -> PipeWriter {
+        PipeWriter { io: io }
```
Comment (Thomasdezeeuw) [inline src/deprecated/redox.rs]: `PipeWriter { io }`

**Comment by jD91mZM2** [comment]
Could you restart the appveyor test? I really doubt I affected that :stuck_out_tongue:

File: src/net/tcp.rs (hunk 34 above)
[the remark this answers is not in the document]
Comment (Thomasdezeeuw) [reply src/net/tcp.rs]: Your right I was look at `TcpStream::connect`, but it uses `SocketAddr::V4`.

File: src/sys/redox/selector.rs (hunk 20 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/selector.rs]: Ok, nevermind then.

File: src/sys/redox/selector.rs (hunk 21 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/selector.rs]: Didn't about `File::open` allocating. Then lets leave this for a future pr.

File: src/sys/redox/eventedfd.rs (hunk 6 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/eventedfd.rs]: Maybe crate a new file `src/sys/eventedfd.rs` with just the struct definition, then only use it with unix and redox. And implement `FromRawFd` etc. separately in the redox and unix modules.

File: src/sys/redox/mod.rs (hunk 18 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/mod.rs]: Ah, didn't know that.

File: src/sys/redox/mod.rs (hunk 15 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/mod.rs]: Well I would prefer it to be a bit clearer, but that can be left for a future pr then.

File: src/sys/redox/selector.rs (hunk 23 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/selector.rs]: Could you add a note about that? Now it seems like the code is just left in by accident.

File: src/sys/redox/udp.rs (hunk 30 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/udp.rs]: Ok. So I would be correct to assum it's not implemented in Redox?

File: src/sys/redox/selector.rs (hunk 27 above)
Comment (Thomasdezeeuw) [reply src/sys/redox/selector.rs]: Ok, well then it's fine to leave it.

**Comment by jD91mZM2** [comment]
I'd love to add CI support, but the policy for adding new platforms has not yet been completely resolved, and @asomers' suggested changes would imply that redox starts off without a CI until it gets a lower tier.

**Comment by carllerche** [comment]
At the very least, it would be nice if there was some sort of cross compilation test to ensure that it compiled at the very least.

**Comment by carllerche** [comment]
@hawkw could you review this? I would not focus on any of the implementation details. The only thing of import is that no existing code is impacted short of mod declarations & `use` statements that are scoped with a cfg for redox.

File: src/sys/unix/tcp.rs [hunk 40]
```diff
@@ -54,11 +54,7 @@ impl TcpStream {
     }
 
     pub fn try_clone(&self) -> io::Result<TcpStream> {
-        self.inner.try_clone().map(|s| {
-            TcpStream {
-                inner: s,
-            }
-        })
+        self.inner.try_clone().map(Self::from_stream)
```
Comment (hawkw) [inline src/sys/unix/tcp.rs]: This change seems unrelated from the rest of this branch?

File: src/net/tcp.rs [hunk 41]
```diff
@@ -89,18 +93,26 @@ impl TcpStream {
...
         let sock = match *addr {
             SocketAddr::V4(..) => TcpBuilder::new_v4(),
             SocketAddr::V6(..) => TcpBuilder::new_v6(),
         }?;
         // Required on Windows for a future `connect_overlapped` operation to be
         // executed successfully.
-        if cfg!(windows) {
-            sock.bind(&inaddr_any(addr))?;
-        }
+        #[cfg(windows)]
```
Comment (hawkw) [inline src/net/tcp.rs]: The change from `if cfg!(windows) { ... }` to `#[cfg(windows)]` is probably better, but seems unrelated to adding Redox support.

File: src/net/tcp.rs [hunk 42]
```diff
@@ -389,6 +401,7 @@ impl TcpStream {
     }
 }
 
+#[cfg(windows)]
```
Comment (hawkw) [inline src/net/tcp.rs]: Again, this seems unrelated to Redox support.

File: src/net/tcp.rs [hunk 43]
```diff
@@ -523,9 +537,8 @@ impl TcpListener {
         }?;
 
         // Set SO_REUSEADDR, but only on Unix (mirrors what libstd does)
-        if cfg!(unix) {
-            sock.reuse_address(true)?;
-        }
+        #[cfg(unix)]
```
Comment (hawkw) [inline src/net/tcp.rs]: Change from `if cfg!(...)` to `#[cfg(...)]` attribute seems unrelated.

File: src/net/tcp.rs (hunk 34 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply src/net/tcp.rs]: Looks like an unrelated change?

File: Cargo.toml [hunk 44]
```diff
@@ -26,18 +26,21 @@ with-deprecated = []
...
 [target.'cfg(target_os = "fuchsia")'.dependencies]
 fuchsia-zircon = "0.3.2"
 fuchsia-zircon-sys = "0.3.2"
 
+[target.'cfg(target_os = "redox")'.dependencies]
+redox_syscall = "0.1"
+
 [target.'cfg(unix)'.dependencies]
-libc   = "0.2.42"
+libc = "0.2.42"
```
Comment (hawkw) [inline Cargo.toml]: Unrelated formatting change (though, this did need to be fixed, I think).

File: Cargo.toml [hunk 45]
```diff
@@ -26,18 +26,21 @@ with-deprecated = []
 default = ["with-deprecated"]
 
 [dependencies]
+iovec    = "0.1.2"
 lazycell = "1"
 log      = "0.4"
+net2     = "0.2.33"
```
Comment (hawkw) [inline Cargo.toml]: Were the dependency version bumps necessary for adding Redox support?

Comment (jD91mZM2) [reply Cargo.toml]: Yes, net2 only recently has redox support

**Comment by jD91mZM2** [comment]
@carllerche Didn't want to set up a CI until I got thumbs up, especially with @asomers' proposed platform policy (which would only allow a CI at tier 2). With your permission I shall do so now.

**Comment by jD91mZM2** [comment]
Guessing the current CI issue is unrelated?

File: src/lib.rs [hunk 46]
```diff
@@ -198,12 +201,13 @@ pub use poll::Iter as EventsIter;
 #[doc(hidden)]
 pub use io::deprecated::would_block;
 
-#[cfg(all(unix, not(target_os = "fuchsia")))]
+#[cfg(any(all(unix, not(target_os = "fuchsia")), target_os = "redox"))]
```
Comment (carllerche) [inline src/lib.rs]: Is it possible to move this to a separate block? So, split out the `cfg(redox)` and have a separate `pub mod unix`? It would make it easier to manage IMO.

File: src/net/tcp.rs [hunk 47]
```diff
@@ -7,12 +7,16 @@
 //!
 /// [portability guidelines]: ../struct.Poll.html#portability
 
+#[cfg(not(target_os = "redox"))]
```
Comment (carllerche) [inline src/net/tcp.rs]: Could you explain on why this file needs changes?

I would prefer to avoid adding platform cfg guards in `src/net/tcp`.

File: src/net/tcp.rs [hunk 48]
```diff
@@ -683,24 +709,24 @@ impl fmt::Debug for TcpListener {
  *
  */
 
-#[cfg(all(unix, not(target_os = "fuchsia")))]
+#[cfg(any(all(unix, not(target_os = "fuchsia")), target_os = "redox"))]
```
Comment (carllerche) [inline src/net/tcp.rs]: I'm wondering if these impls should be moved out of this file in order to maintain the goal of "no cfgs" in this file.

File: src/net/tcp.rs (hunk 47 above)
Comment (jD91mZM2) [reply src/net/tcp.rs]: Back when I wrote this, I would argue that "redox doesn't support creating non-blocking sockets before an address is bound". I am extremely certain that I tried it, but... I've seen the source code of redox' libc now, and it seems to disagree.

It's still always better to use native rust where possible (redox' backend doesn't use libc), but I cannot safely say the old code won't wory. I'll try it again, later.

Comment (jD91mZM2) [reply src/net/tcp.rs]: Just double checked, also checked the code. There *are* null sockets implemented, but they don't support fcntl. I'll ask Jeremy what he'd like to do here.

**Comment by jD91mZM2** [comment]
The latest changes separate stuff into their respective platforms. This obviously also affects unix/windows, but not more than strictly necessary. I might be able to revert and keep redox consistent this if the redox netstack team gives me thumbs up to implement fcntl for unbound sockets, but then again that depends on libc so this is better anyway.

**Comment by jD91mZM2** [comment]
Whatever I said about fcntl depending on libc was stupid because it does the direct system call and in no way does using that depend on another system call. And as I need unbound sockets for openssl anyway, I'm implementing that right now.

**Comment by jD91mZM2** [comment]
There we go, unbound fcntls, no redox-specific code for connecting. This *does* mean it uses libc for making the connection (net2), so tell me if you want me to resolve this in some other way.

**Comment by jD91mZM2** [comment]
BUMP?

**Comment by carllerche** [comment]
Ok, here is the plan. Lets get this in the 0.7 release of Mio.

I commented with a proposed implementation strategy for adding new platforms [here](https://github.com/tokio-rs/mio/issues/853#issuecomment-490630832). This proposal is based on how this PR needs to touch so many files. I want to avoid this.

So, if my proposed strategy is accepted, the contents of `sys/redox` would be moved into a dedicated crate that mio would conditionally depend on (`repoll`?). Also, we would split out the cfg flags... I'll probably work on some of this restructuring on master.

wdyt?

**Comment by jD91mZM2** [comment]
I appreciate the idea, it sounds pretty great! I'm afraid, however, I might just drop this PR in its entirety as https://github.com/rust-lang/rust/issues/60139 is merging. Redox is moving its focus away from updating the world to rely directly on Redox APIs, and instead making the C API cover everything. So I am sorry if I made you figure out this policy for smaller gain than expected, but I hope you can still use it for the other platforms as it sounds like a great idea in general.

**Comment by carllerche** [comment]
Sorry about all that. The policy is still going to be valuable I think.

However, even w/ redox targettng `unix`, I would guess that some level of redox specific integration is going to be needed. There are linux, bsd, etc... specific code for implementing the selector.

**Comment by jD91mZM2** [comment]
That's true, someone will probably open a small PR to add redox to the linux `cfg`s. We seem to be going the `epoll` route in our C standard library relibc. :smile:

**Comment by Thomasdezeeuw** [comment]
@jD91mZM2 this is likely not the place for it, but I would recommend `kqueue` over `epoll`. The API in my opinion is much easier to use and supports I/O, signals, timers and use space events all with 2 system calls.

**Comment by jD91mZM2** [comment]
@Thomasdezeeuw I will look up `kqueue` and hear with the BDFL if he thinks implementing it instead (or alongside?) of his `epoll` implementation is worth it. Thanks a lot!
